File: python/1034. Coloring A Border.py
import copy

class Solution(object):
    def colorBorder(self, grid, r0, c0, color):
        """
        :type grid: List[List[int]]
        :type r0: int
        :type c0: int
        :type color: int
        :rtype: List[List[int]]
        """
        m = len(grid)
        n = len(grid[0])
        vis = [[0]*n for _ in range(m)]
        mp = copy.deepcopy(grid)

        def dfs(x, y, c):
            if x<0 or x>m-1 or y<0 or y>n-1 or mp[x][y]!=c:
                return
            vis[x][y] = 1
            for dx, dy in [(1, 0), (-1, 0), (0, 1), (0, -1)]:
                nx = x + dx
                ny = y + dy
                
                if nx<0 or nx>m-1 or ny<0 or ny>n-1 or mp[nx][ny]!=c:
                    grid[x][y]=color
                else:
                    if vis[nx][ny]:
                        continue
                    dfs(nx, ny, c)
                

        dfs(r0, c0, grid[r0][c0])
        return grid


if __name__ == '__main__':
    grid = [[1,1,1],[1,1,1],[1,1,1]]
    r0 = 1
    c0 = 1
    color = 2
    su = Solution()
    print su.colorBorder(grid, r0, c0, color)        

File: go/src/leetcode/1034. Coloring A Border/1034. Coloring A Border.go
/**
@author ZhengHao Lou
Date    2021/12/07
*/
package main

import "fmt"

/**
https://leetcode-cn.com/problems/coloring-a-border/
 */
var offset = [][]int{{1,0},{0,-1},{-1,0},{0,1}}
func colorBorder(grid [][]int, row int, col int, color int) [][]int {
	m, n := len(grid), len(grid[0])
	c := grid[row][col]
	vis := make([][]bool, m)
	for i := range vis {
		vis[i] = make([]bool, n)
	}
	que := []int{row*n + col}
	for len(que) != 0 {
		i := que[0]
		que = que[1:]
		x, y := i/n, i%n
		vis[x][y] = true
		var edge = x == m-1 || y == n-1 || x == 0 || y == 0
		for _, dxy := range offset {
			nx, ny := x + dxy[0], y + dxy[1]
			if nx < 0 || nx >= m || ny < 0 || ny >= n || vis[nx][ny] {
				continue
			}
			if grid[nx][ny] != c {
				edge = true
			} else {
				que = append(que,  nx*n + ny)
			}
		}

		if edge {
			grid[x][y] = color
		}
	}

	fmt.Println(grid)
	return grid
}

func main() {
	colorBorder([][]int{{1,1},{1,2}}, 0, 0, 3)
	colorBorder([][]int{{1,2,2},{2,3,2}}, 0, 1, 3)
	colorBorder([][]int{{1,1,1},{1,1,1},{1,1,1}}, 1, 1, 2)
}
